File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        m = len(board)
        n = len(board[0])

        def dfs(r,c, idx):
            if idx == len(word):
                return True            
            if r>=m or c >= n or r < 0 or c < 0 or board[r][c] != word[idx]:
                return False

            temp, board[r][c] = board[r][c], '/'
            res = dfs(r+1, c, idx+1) or dfs(r-1, c, idx+1) or dfs(r, c+1, idx+1) or dfs(r, c-1, idx+1)
            board[r][c] = temp

            return res

        for r in range(m):
            for c in range(n):
                if board[r][c] == word[0]:
                    if dfs(r,c,0) == True:
                        return True
        return False

File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        if(word.size() ==0){
            return true;
        }
        if(board.size()==0){
            return false;
        }
        for(int i=0; i<board.size(); i++){
            for(int j=0; j<board[0].size(); j++){
                if(board[i][j] == word[0]){
                    if(dfs(i,j,0,board,word)){
                        return true;
                    }
                }
            }
        }
        return false;
    }
    bool dfs(int r, int c, int idx, vector<vector<char>>& board, string word){
        if(idx == word.size()) {
            return true;
        }
        if(r>=board.size() || r<0 || c>=board[0].size() || c<0 || word[idx]!=board[r][c]) 
        {
            return false;
        }
        char temp = board[r][c];
        board[r][c] = '#';
        
        bool found = dfs(r+1,c,idx+1,board,word)||
                     dfs(r-1,c,idx+1,board,word)||
                     dfs(r,c+1,idx+1,board,word)||
                     dfs(r,c-1,idx+1,board,word);
        board[r][c] = temp;
        return found;
    }
};